export_elasticsearch doesn't do Array[Struct{..}] => nested object

### Hail version:
0.1 only - this already works in 0.2

### What you did:
I'm calling kt.export_elasticsearch(..) on this keytable schema:

```
     ...
     aIndex: Int,
     alt: String,
     codingGeneIds: Set[String],
     contig: String,
     docId: String,
     domains: Set[String],
     end: Int,
     filters: Set[String],
     geneIds: Set[String],
     pos: Int,
     ref: String,
     rsid: String,
     start: Int,
     variantId: String,
     vep: Array[Struct{
         gene_id: String,
         gene_symbol: String,
         protein_id: String,
         transcript_id: String,
         hgvs: String,
         major_consequence: String,
         major_consequence_rank: Int,
         category: String
     }],
     wasSplit: Boolean,
```

### What went wrong (all error messages here, including the full java stack trace):

Having `vep: Array[Struct..]` in the schema leads to 

```
2018-09-02 07:06:06,821 INFO     ==> exporting data to elasticasearch. Write mode: upsert, blocksize: 1000
Config Map(es.batch.size.entries -> 1000, es.index.auto.create -> true, es.mapping.id -> docId, es.write.operation -> upsert, es.port -> 9200, es.nodes -> 10.56.10.4)
[Stage 3:>                                                     (0 + 100) / 1000]Traceback (most recent call last):
  File "/tmp/ffc9fb0b99f64080b674ab7a07962df9/load_dataset_to_es.py", line 734, in <module>
    run_pipeline()
  File "/tmp/ffc9fb0b99f64080b674ab7a07962df9/load_dataset_to_es.py", line 726, in run_pipeline
    hc, vds = step2_export_to_elasticsearch(hc, vds, args)
  File "/tmp/ffc9fb0b99f64080b674ab7a07962df9/load_dataset_to_es.py", line 565, in step2_export_to_elasticsearch
    disable_index_for_fields=("sortedTranscriptConsequences", ),
  File "/tmp/ffc9fb0b99f64080b674ab7a07962df9/load_dataset_to_es.py", line 358, in export_to_elasticsearch
    verbose=True,
  File "/tmp/ffc9fb0b99f64080b674ab7a07962df9/hail_scripts.zip/hail_scripts/v01/utils/elasticsearch_client.py", line 140, in export_vds_to_elasticsearch
  File "/tmp/ffc9fb0b99f64080b674ab7a07962df9/hail_scripts.zip/hail_scripts/v01/utils/elasticsearch_client.py", line 287, in export_kt_to_elasticsearch
  File "<decorator-gen-143>", line 2, in export_elasticsearch
  File "/tmp/ffc9fb0b99f64080b674ab7a07962df9/hail-0.1-es-6.2.4-with-strip-chr-prefix.zip/hail/java.py", line 121, in handle_py4j
hail.java.FatalError: EsHadoopIllegalArgumentException: Spark SQL types are not handled through basic RDD saveToEs() calls; typically this is a mistake(as the SQL schema will be ignored). Use 'org.elasticsearch.spark.sql' package instead

Java stack trace:
org.apache.spark.SparkException: Job aborted due to stage failure: Task 68 in stage 3.0 failed 20 times, most recent failure: Lost task 68.19 in stage 3.0 (TID 3771, vep-grch37-sw-9767.c.seqr-project.internal): org.elasticsearch.hadoop.EsHadoopIllegalArgumentException: Spark SQL types are not handled through basic RDD saveToEs() calls; typically this is a mistake(as the SQL schema will be ignored). Use 'org.elasticsearch.spark.sql' package instead
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWriteScala(ScalaValueWriter.scala:124)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWrite(ScalaValueWriter.scala:50)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$3.apply(ScalaValueWriter.scala:78)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$3.apply(ScalaValueWriter.scala:77)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWriteScala(ScalaValueWriter.scala:77)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWrite(ScalaValueWriter.scala:50)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$2.apply(ScalaValueWriter.scala:66)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$2.apply(ScalaValueWriter.scala:63)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:221)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWriteScala(ScalaValueWriter.scala:63)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.write(ScalaValueWriter.scala:46)
	at org.elasticsearch.hadoop.serialization.builder.ContentBuilder.value(ContentBuilder.java:53)
	at org.elasticsearch.hadoop.serialization.bulk.TemplatedBulk.doWriteObject(TemplatedBulk.java:71)
	at org.elasticsearch.hadoop.serialization.bulk.TemplatedBulk.write(TemplatedBulk.java:58)
	at org.elasticsearch.hadoop.rest.RestRepository.writeToIndex(RestRepository.java:168)
	at org.elasticsearch.spark.rdd.EsRDDWriter.write(EsRDDWriter.scala:67)
	at org.elasticsearch.spark.rdd.EsSpark$$anonfun$doSaveToEs$1.apply(EsSpark.scala:107)
	at org.elasticsearch.spark.rdd.EsSpark$$anonfun$doSaveToEs$1.apply(EsSpark.scala:107)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
	at org.apache.spark.scheduler.Task.run(Task.scala:86)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1454)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1442)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1441)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1441)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:811)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:811)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:811)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1667)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1622)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1611)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:632)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1873)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1886)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1906)
	at org.elasticsearch.spark.rdd.EsSpark$.doSaveToEs(EsSpark.scala:107)
	at org.elasticsearch.spark.rdd.EsSpark$.saveToEs(EsSpark.scala:79)
	at org.elasticsearch.spark.rdd.EsSpark$.saveToEs(EsSpark.scala:76)
	at is.hail.io.ElasticsearchConnector$.export(ElasticsearchConnector.scala:33)
	at is.hail.keytable.KeyTable.exportElasticsearch(KeyTable.scala:751)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:237)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)org.elasticsearch.hadoop.EsHadoopIllegalArgumentException: Spark SQL types are not handled through basic RDD saveToEs() calls; typically this is a mistake(as the SQL schema will be ignored). Use 'org.elasticsearch.spark.sql' package instead
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWriteScala(ScalaValueWriter.scala:124)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWrite(ScalaValueWriter.scala:50)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$3.apply(ScalaValueWriter.scala:78)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$3.apply(ScalaValueWriter.scala:77)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWriteScala(ScalaValueWriter.scala:77)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWrite(ScalaValueWriter.scala:50)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$2.apply(ScalaValueWriter.scala:66)
	at org.elasticsearch.spark.serialization.ScalaValueWriter$$anonfun$doWriteScala$2.apply(ScalaValueWriter.scala:63)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:221)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.doWriteScala(ScalaValueWriter.scala:63)
	at org.elasticsearch.spark.serialization.ScalaValueWriter.write(ScalaValueWriter.scala:46)
	at org.elasticsearch.hadoop.serialization.builder.ContentBuilder.value(ContentBuilder.java:53)
	at org.elasticsearch.hadoop.serialization.bulk.TemplatedBulk.doWriteObject(TemplatedBulk.java:71)
	at org.elasticsearch.hadoop.serialization.bulk.TemplatedBulk.write(TemplatedBulk.java:58)
	at org.elasticsearch.hadoop.rest.RestRepository.writeToIndex(RestRepository.java:168)
	at org.elasticsearch.spark.rdd.EsRDDWriter.write(EsRDDWriter.scala:67)
	at org.elasticsearch.spark.rdd.EsSpark$$anonfun$doSaveToEs$1.apply(EsSpark.scala:107)
	at org.elasticsearch.spark.rdd.EsSpark$$anonfun$doSaveToEs$1.apply(EsSpark.scala:107)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
	at org.apache.spark.scheduler.Task.run(Task.scala:86)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)


Hail version: 0.1-408f188
Error summary: EsHadoopIllegalArgumentException: Spark SQL types are not handled through basic RDD saveToEs() calls; typically this is a mistake(as the SQL schema will be ignored). Use 'org.elasticsearch.spark.sql' package instead
ERROR: (gcloud.dataproc.jobs.submit.pyspark) Job [ffc9fb0b99f64080b674ab7a07962df9] entered state [ERROR] while waiting for [DONE].
```

Ideally it would get exported as nested objects: https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html#_using_literal_nested_literal_fields_for_arrays_of_objects

with elasticsearch mapping:
```
u'vep': {'type': 'nested', 'properties': {u'category': {'type': 'keyword'}, u'major_consequence': {'type': 'keyword'}, u'gene_id': {'type': 'keyword'}, u'major_consequence_rank': {'type': 'integer'}, u'gene_symbol': {'type': 'keyword'}, u'transcript_id': {'type': 'keyword'}, u'hgvs': {'type': 'keyword'}, u'protein_id': {'type': 'keyword'}}}, 
```

I thought about switching to saveJsonToEs here: 
https://github.com/hail-is/hail/blob/0.1/src/main/scala/is/hail/io/ElasticsearchConnector.scala#L33
but not sure how to auto-convert to json before exporting
 https://www.elastic.co/guide/en/elasticsearch/hadoop/6.x/spark.html#spark-streaming-write-json

--- a/src/main/scala/is/hail/io/ElasticsearchConnector.scala
+++ b/src/main/scala/is/hail/io/ElasticsearchConnector.scala
@@ -1,10 +1,8 @@
 package is.hail.io
 
-import is.hail.HailContext
 import is.hail.keytable.KeyTable
 import org.elasticsearch.spark.sql._
 import collection.JavaConverters._
-import scala.collection.mutable
 
 import scala.collection.Map
 

--- a/src/main/scala/is/hail/io/ElasticsearchConnector.scala
+++ b/src/main/scala/is/hail/io/ElasticsearchConnector.scala
@@ -1,11 +1,27 @@
 package is.hail.io
 
+import is.hail.HailContext
 import is.hail.keytable.KeyTable
-import org.elasticsearch.spark.rdd.EsSpark
-import scala.collection.Map
+import org.elasticsearch.spark.sql._
+import collection.JavaConverters._
 import scala.collection.mutable
 
+import scala.collection.Map
+
 object ElasticsearchConnector {
+
+  def export(
+              kt: KeyTable,
+              host: String,
+              port: Int,
+              index: String,
+              indexType: String,
+              blockSize: Int,
+              config: java.util.HashMap[String, String],
+              verbose: Boolean) {
+    export(kt, host, port, index, indexType, blockSize,
+      Option(config).map(_.asScala.toMap).getOrElse(Map.empty[String, String]), verbose)
+  }
 
   def export(kt: KeyTable, host: String = "localhost", port: Int = 9200,
              index: String, indexType: String, blockSize: Int = 1000,
@@ -17,24 +33,19 @@ object ElasticsearchConnector {
       "es.nodes" -> host,
       "es.port" -> port.toString,
       "es.batch.size.entries" -> blockSize.toString,
-      "es.index.auto.create" -> "true"
-    )
+      "es.index.auto.create" -> "true")
 
     val mergedConfig = if (config == null)
-        defaultConfig
-      else
-        defaultConfig ++ config
+      defaultConfig
+    else
+      defaultConfig ++ config
 
     if (verbose)
-      println(s"Config ${mergedConfig}")
+      println(s"Config ${ mergedConfig }")
 
-    val fields = kt.signature.fields.map(_.name)
-
-    EsSpark.saveToEs(
-      kt.rdd.map(r => fields.zip(r.toSeq).toMap[String, Any]),
-      s"${index}/${indexType}",
-      mergedConfig
-    )
-
+    val df = kt
+      .expandTypes()
+      .toDF(kt.hc.sqlContext)
+      .saveToEs(s"${ index }/${ indexType }", mergedConfig)
   }
 }

